### ISSUE TITLE
How to do a OpenMP parallel for loop over gf_mesh<imfreq>?

### ISSUE THREAD

**Issue by HugoStrand** [TRIQS/triqs#560]
Dear all,

I am trying to make a little OpenMP parallelization of a loop over a `gf_mesh<imfreq>` mesh. The original code uses the nice `c++17` syntax
```c++
for (auto const &k : kmesh) { ... }
```
Since OpenMP (4.0) does not yet support this range-based for loop syntax I have tried to use the iterator interface directly, i.e.
```c++
#pragma omp parallel for
for (auto k_iter = kmesh.begin(); k_iter != kmesh.end(); k_iter++ ) { 
  auto k = *k_iter;
  ... 
}
```
This does not compile since OpenMP requires the comparison operator to be `<` or `<=`. However that is not implemented by the iterator interface of `gf_mesh<imfreq>`. Trying with the `<` operator I get the error message:
```
error: invalid operands to binary expression
      ('triqs::gfs::mesh_pt_generator<triqs::gfs::cluster_mesh>' and 'cluster_mesh::const_iterator' (aka
      'mesh_pt_generator<triqs::gfs::cluster_mesh>'))
    for (auto q_iter = kmesh.begin(); q_iter < kmesh.end(); q_iter++ ) {
                                      ~~~~~~ ^ ~~~~~~~~~~~
/Users/hstrand/apps/triqs_unstable/include/triqs/./clef/adapters/../clef.hpp:205:2: note: candidate template ignored:
      requirement 'is_any_lazy<mesh_pt_generator<cluster_mesh> &, mesh_pt_generator<cluster_mesh> >::value' was not
      satisfied [with L = triqs::gfs::mesh_pt_generator<triqs::gfs::cluster_mesh> &, R =
      triqs::gfs::mesh_pt_generator<triqs::gfs::cluster_mesh>]
 TRIQS_CLEF_OPERATION(less,       <);
 ^
/Users/hstrand/apps/triqs_unstable/include/triqs/./clef/adapters/../clef.hpp:191:114: note: expanded from macro
      'TRIQS_CLEF_OPERATION'
  ...std14::enable_if_t<is_any_lazy<L, R>::value, expr<tags::TAG, expr_storage_t<L>, expr_storage_t<R>>> operator OP(L&& ...
                                                                                                         ^
/Users/hstrand/apps/triqs_unstable/include/triqs/arrays/./expression_template/../linalg/../vector.hpp:256:84: note: 
      candidate template ignored: requirement 'ImmutableVector<mesh_pt_generator<cluster_mesh> >::value' was not satisfied
      [with V1 = triqs::gfs::mesh_pt_generator<triqs::gfs::cluster_mesh>, V2 =
      triqs::gfs::mesh_pt_generator<triqs::gfs::cluster_mesh>]
 std14::enable_if_t<ImmutableVector<V1>::value&& ImmutableVector<V2>::value, bool> operator<(V1 const& a, V2 const& b) {
                                                  
```

Can we fix this somehow?

Best, Hugo

**Comment by HugoStrand** [comment]
Dear all,

Here is a poor man's solution for an OpenMP parallelized k-mesh loop.

```c++
#pragma omp parallel for 
    for (int qidx = 0; qidx < kmesh.size(); qidx++) {
      auto q_iter = kmesh.begin();
      q_iter += qidx;
      auto q = *q_iter;
    ....
}
```

**Comment by HugoStrand** [comment]
After discussions with @Wentzell we concluded that the long term solution is to change the mesh iterators from forward iterators to random access iterators. This will enable more straight forward OpenMP parallelization.

Please, keep this thread until this has been implemented or the request is discarded.

Best, Hugo

**Comment by krivenko** [comment]
@HugoStrand This short discussion could be of relevance:  #131.

**Comment by parcollet** [comment]
- Does the open MP specs require RandomAccessIterator ??

- Can we in all cases make the iterator random access efficiently ?

**Comment by Wentzell** [comment]
This is fixed by #691
